From ae8df9cc98e900fa40883c525eef5282036c32fb Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Wed, 15 Oct 2025 17:26:36 +0200
Subject: [PATCH 432/553] QApplication: reset WA_KeyboardFocusChange on focus
 change

WA_KeyboardFocusChange is set when the user changes the focus by mouse
but never cleared once the focus was changed e.g. with the mouse. The
only way to clear it is to unfocus the whole top-level widget. This is
results in focus rects in some style like fusion on checkboxes where
there shouldn't be one.
Fix it by clearing the attribute when the FocusReason is not keyboard.

Fixes: QTBUG-141100
Change-Id: I211e7f435c9fe748ee9925e35b481166b7446a55
Reviewed-by: Oliver Wolff <oliver.wolff@qt.io>
Reviewed-by: Friedemann Kleint <Friedemann.Kleint@qt.io>
(cherry picked from commit 6cd906a8bd80b62d254c9491720909cd87cfa36b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/kernel/qapplication.cpp | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/src/widgets/kernel/qapplication.cpp b/src/widgets/kernel/qapplication.cpp
index dd91e14a257..6ff5ce55bf5 100644
--- a/src/widgets/kernel/qapplication.cpp
+++ b/src/widgets/kernel/qapplication.cpp
@@ -1518,11 +1518,15 @@ void QApplicationPrivate::setFocusWidget(QWidget *focus, Qt::FocusReason reason)
             return;
         }
 
-        if (focus && (reason == Qt::BacktabFocusReason || reason == Qt::TabFocusReason)
-            && qt_in_tab_key_event)
-            focus->window()->setAttribute(Qt::WA_KeyboardFocusChange);
-        else if (focus && reason == Qt::ShortcutFocusReason) {
-            focus->window()->setAttribute(Qt::WA_KeyboardFocusChange);
+        if (focus) {
+            if ((reason == Qt::BacktabFocusReason || reason == Qt::TabFocusReason)
+                 && qt_in_tab_key_event)
+                focus->window()->setAttribute(Qt::WA_KeyboardFocusChange);
+            else if (reason == Qt::ShortcutFocusReason) {
+                focus->window()->setAttribute(Qt::WA_KeyboardFocusChange);
+            } else {
+                focus->window()->setAttribute(Qt::WA_KeyboardFocusChange, false);
+            }
         }
         QWidget *prev = focus_widget;
         focus_widget = focus;
-- 
2.51.2

